### PR TITLE
Display layout alphanumeric

### PR DESCRIPTION
**Pull request by psarram**
Support adding multiple telemetry points of the same type to display layouts and use the toolbar to edit their properties.

### CODE REVIEW

File: src/plugins/displayLayout/DisplayLayoutToolbar.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                        },
+                        {
+                            control: "separator"
+                        },
+                        {
+                            control: "select-menu",
+                            domainObject: domainObject,
+                            property: path + ".value",
+                            title: "Set value",
+                            options: metadata.values().map(value => {
```
Comment (evenstensberg) [inline src/plugins/displayLayout/DisplayLayoutToolbar.js]: If metadata is with void values, maybe a filter would be better?

File: src/plugins/displayLayout/DisplayLayoutToolbar.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                        },
+                        {
+                            control: "separator"
+                        },
+                        {
+                            control: "select-menu",
+                            domainObject: domainObject,
+                            property: path + ".size",
+                            title: "Set text size",
+                            options: [9, 10, 11, 12, 13, 14, 15, 16, 20, 24, 30, 36, 48, 72, 96].map(function (size) {
```
Comment (evenstensberg) [inline src/plugins/displayLayout/DisplayLayoutToolbar.js]: This is kinda long, perhaps a `CONST_VAR` would be favorable? Also the map from earlier didn't use `function(val)` signature on its cb, perhaps a consistency there?

File: src/plugins/displayLayout/components/LayoutItem.vue [hunk 3]
```diff
@@ -164,7 +98,7 @@
                     $event.stopPropagation();
                 }
 
-                if (!this.isBeingEdited(this.item.domainObject)) {
+                if (!this.openmct.editor.isEditing()) {
```
Comment (evenstensberg) [inline src/plugins/displayLayout/components/LayoutItem.vue]: nice

File: src/plugins/displayLayout/DisplayLayoutToolbar.js (hunk 1 above)
Comment (larkin) [reply src/plugins/displayLayout/DisplayLayoutToolbar.js]: metadata is a TelemetryMetadataManager not an Array.

File: src/plugins/displayLayout/DisplayLayoutToolbar.js (hunk 2 above)
Comment (psarram) [reply src/plugins/displayLayout/DisplayLayoutToolbar.js]: I will update to use the arrow function here for consistency. Generally it's a better practice to define it as a constant regardless of the length.

File: src/plugins/displayLayout/DisplayLayoutToolbar.js (hunk 1 above)
Comment (psarram) [reply src/plugins/displayLayout/DisplayLayoutToolbar.js]: @evenstensberg all the values that are returned by values() should be used to create the name value pair and none should be filtered.

File: src/plugins/displayLayout/components/DisplayLayout.vue [hunk 4]
```diff
@@ -99,86 +102,96 @@
...
         methods: {
-            readLayoutConfiguration(domainObject, panels) {
-                let id = this.openmct.objects.makeKeyString(domainObject.identifier);
-                this.rawPositions[id] = {
-                    position: panels[id].position || DEFAULT_POSITION,
-                    dimensions: panels[id].dimensions || this.defaultDimensions()
+            getPanels() {
+                let panels = this.newDomainObject.configuration.panels;
+
+                for (const id in panels) {
```
Comment (larkin) [inline src/plugins/displayLayout/components/DisplayLayout.vue]: to use `for ... in` you need to check each id using `hasOwnProperty`.  But thanks to modern JS, you should use `for ... of` with Object.entries in order to iterate over key/value pairs, e.g. `for (const [id, panel] of Object.entries(panels)) { ... }`.

File: src/plugins/displayLayout/components/TelemetryView.vue [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+ <script>
+    export default {
+        inject: ['openmct'],
+        props: {
+            item: Object
+        },
+        computed: {
+            showLabel: function () {
+                let displayMode = this.item.config.alphanumeric.displayMode;
+                return (displayMode === 'all' || displayMode === 'label') ? true : false;
```
Comment (larkin) [inline src/plugins/displayLayout/components/TelemetryView.vue]: don't need the ternary operator when the return value is already a boolean.

File: src/plugins/displayLayout/components/TelemetryView.vue [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            },
+            showValue: function () {
+                let displayMode = this.item.config.alphanumeric.displayMode;
+                return displayMode === 'value' || displayMode == 'all' ? true : false;
```
Comment (larkin) [inline src/plugins/displayLayout/components/TelemetryView.vue]: don't need the ternary operator when the return value is already a boolean.

File: src/plugins/displayLayout/components/TelemetryView.vue [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        },
+        data() {
+            return {
+                telemetryValue: '',
+                telemetryClass: ''
+            }
+        },
+        methods: {
+            getTelemetry(domainObject) {
+                return Promise.resolve(domainObject)
```
Comment (larkin) [inline src/plugins/displayLayout/components/TelemetryView.vue]: This doesn't need to be wrapped in a promise chain, can just be done synchronously.

File: src/plugins/displayLayout/components/TelemetryView.vue [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+                    if (self.openmct.time.clock() !== undefined) {
+                        self.updateView(object, datum);
+                    }
+                }, {});
+
+                return object;
+            },
+            updateView(telemetryObject, datum) {
+                let metadata = this.openmct.telemetry.getMetadata(telemetryObject);
+                let valueMetadata = metadata.values().filter(function (value) {
```
Comment (larkin) [inline src/plugins/displayLayout/components/TelemetryView.vue]: 1. can get metadata for value key with `metadata.value(key)`
2. should cache formatter and limit evaluator instead of re-evaluating this logic for every new data point that comes in.

File: src/ui/components/toolbar/components/toolbar-color-picker.vue [hunk 9]
```diff
@@ -14,7 +14,7 @@
                 v-if="!this.options.preventNone"
                 @click="select({value: 'transparent'})">
                 <div class="c-palette__item"></div>
-                No fill
```
Comment (larkin) [inline src/ui/components/toolbar/components/toolbar-color-picker.vue]: IMO, The `options.type` property is exposing customization that is out of scope for our toolbar components.  This should use a generic label (i.e. `transparent` or `none`) and the type property should be removed from the color picker options.

File: src/plugins/displayLayout/components/TelemetryView.vue (hunk 5 above)
Comment (psarram) [reply src/plugins/displayLayout/components/TelemetryView.vue]: Good point. Will fix it.

File: src/plugins/displayLayout/components/TelemetryView.vue (hunk 7 above)
Comment (psarram) [reply src/plugins/displayLayout/components/TelemetryView.vue]: Will do.

File: src/ui/components/toolbar/components/toolbar-color-picker.vue (hunk 9 above)
Comment (psarram) [reply src/ui/components/toolbar/components/toolbar-color-picker.vue]: If we don't want to expose that customization at that level, I will remove the type property and use "None" as a generic label.

Comment (larkin) [reply src/ui/components/toolbar/components/toolbar-color-picker.vue]: Yes, please.

File: src/plugins/displayLayout/components/DisplayLayout.vue (hunk 4 above)
Comment (psarram) [reply src/plugins/displayLayout/components/DisplayLayout.vue]: getPanels() is removed and so this code doesn't apply.

File: src/plugins/displayLayout/components/LayoutItem.vue [hunk 10]
```diff
@@ -208,24 +138,30 @@
...
                 document.body.removeEventListener('mouseup', this.endDrag);
                 this.continueDrag(event);
-                this.$emit('endDrag', this.item.id);
+                this.$emit('endDrag', this.item);
                 this.initialPosition = undefined;
                 event.preventDefault();
             }
         },
         mounted() {
+            let context = {};
```
Comment (larkin) [inline src/plugins/displayLayout/components/LayoutItem.vue]: why not
```
let context = {
    item: this.item.domainObject,
    layoutItem = this.item
}
``` 
?

File: src/plugins/displayLayout/components/SubobjectView.vue [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    export default {
+        inject: ['openmct'],
+        props: {
+            item: Object
+        },
+        components: {
+            ObjectView,
+        },
+        mounted() {
+            this.item.config.attachListeners();
```
Comment (larkin) [inline src/plugins/displayLayout/components/SubobjectView.vue]: Hrm, this looks funky but not sure if there's better handling available.  But we can address this next sprint w/ display layout drawing.

File: src/plugins/displayLayout/components/LayoutItem.vue (hunk 10 above)
Comment (psarram) [reply src/plugins/displayLayout/components/LayoutItem.vue]: At some point I tried to build the context conditionally before deciding to pass the layoutItem instead, and it's a left over code. Thanks for pointing it out. It's fixed.

File: src/plugins/displayLayout/components/SubobjectView.vue (hunk 11 above)
Comment (psarram) [reply src/plugins/displayLayout/components/SubobjectView.vue]: The listeners are defined in the view configuration classes for each type (subobject and telemetry). We can discuss offline.

File: src/plugins/displayLayout/components/DisplayLayout.vue [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+                    });
+                }
+            },
+            addAlphanumeric(domainObject, position) {
+                let alphanumeric = {
+                    identifier: domainObject.identifier,
+                    position: position,
+                    dimensions: DEFAULT_TELEMETRY_DIMENSIONS,
+                    displayMode: 'all',
+                    value: 'sin',
```
[the remark this answers is not in the document]
Comment (psarram) [reply src/plugins/displayLayout/components/DisplayLayout.vue]: @larkin I made changes to set the default value in this order:

- If there's a range value, it defaults to the first item in the array.
- If no range is defined, it defaults to the first value that is returned by `metadata.values()` that is non time-domain. 
- Otherwise, it defaults to the first value returned by `metadata.values()`.

Please let me know if anything else needs to be changed. If not, the code is ready to be merged. Thanks for the review!

**Comment by larkin** [comment]
This is looking good!  Going to merge to TCR.
